Material DataTable: added support of setting table row border thickness

dratushnyy opened the pull request:
## Description
At the current moment, there is no option in the `DataTable` widget to setup row thickness since the value of it is hardcoded and set to `1.0` (please refer to the related issue for more detailed description).
This pull request introduced changes fixing the issue, namely, try to use `Theme.of(context).dividerTheme.thickness` and if it is `null` then use a default value of `1.0` which was moved to a named constant inside `DataTable` class to increase code readability.

## Related Issues
Fixes https://github.com/flutter/flutter/issues/49691

## Tests

I added the following tests:
`DataTable set border width test` into `packages\flutter\test\material\data_table_test.dart`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

dratushnyy committed:
Material DataTable: added support of setting table row border thickness
using Theme.dividerTheme.thickness or, if the theme contains null-value
- the value of DataTable._defaultBodrerThickness constant
which was introduced to provide more meaningful value instead
of using nameless double.

Piinks reviewed (commented):
Hi @dratushnyy welcome! 🎉  Thank you for the contribution and for writing a test!
I think this change looks ok, although @HansMuller may have some thoughts here. Particularly regarding www.flutter.dev/go/material-theme-system-updates

HansMuller commented:
For now, it would safer to just add DataTable.dividerThickness, and not to defer to the dividerTheme.  As [Kate mentioned](https://github.com/flutter/flutter/pull/49692#pullrequestreview-356349095), we're moving toward a theme system where there's an overall theme with a small number of properties, and component specific themes.

dratushnyy committed:
Material DataTable: border thickess moved to DataTable
constructor parameter instead of using ThemeData during to
upcoming (and possibly breaking) changes to Material Theme
System.

dratushnyy commented:
@Piinks, @HansMuller  Thank you for your comments, I updated the pull-request accordingly

HansMuller commented on packages/flutter/lib/src/material/data_table.dart:
```diff
@@ -331,6 +331,7 @@ class DataTable extends StatelessWidget {
     this.horizontalMargin = 24.0,
     this.columnSpacing = 56.0,
     this.showCheckboxColumn = true,
+    this.borderThickness = 1.0,
```
I'd prefer to call this property "dividerThickness". Borders typically encircle a widget, dividers fit in between adjacent vertical or horizontal widget edges.

HansMuller commented on packages/flutter/lib/src/material/data_table.dart:
```diff
@@ -343,6 +344,7 @@ class DataTable extends StatelessWidget {
        assert(showCheckboxColumn != null),
        assert(rows != null),
        assert(!rows.any((DataRow row) => row.cells.length != columns.length)),
+       assert(borderThickness != null && borderThickness >=0),
```
` >=0)` should be ` >= 0)`

HansMuller commented on packages/flutter/lib/src/material/data_table.dart:
```diff
@@ -467,6 +469,11 @@ class DataTable extends StatelessWidget {
   static const Color _grey100Opacity = Color(0x0A000000); // Grey 100 as opacity instead of solid color
   static const Color _grey300Opacity = Color(0x1E000000); // Dark theme variant is just a guess.
 
+  /// The width of the border between table rows.
```
```dart
/// The width of the divider that appears between [TableRow]s.
/// 
/// ...
```

HansMuller commented on packages/flutter/lib/src/material/data_table.dart:
```diff
@@ -467,6 +469,11 @@ class DataTable extends StatelessWidget {
   static const Color _grey300Opacity = Color(0x1E000000); // Dark theme variant is just a guess.
 
+  /// The width of the border between table rows.
+  /// Must not be null or negative. 0 is allowed.
```
Must be non-null and greater than or equal to zero.

HansMuller commented on packages/flutter/test/material/data_table_test.dart:
```diff
@@ -866,4 +866,89 @@ void main() {
   });
+
+  testWidgets('DataTable set border width test', (WidgetTester tester) async {
+    Widget buildTable({ double thickness}) {
```
```dart
({ double thickness })
```

HansMuller replied:
The table built by this function seems needlessly complex for the sake of checking the divider.

It might be simpler to just create columns and rows (configured as simply as possible) and then create the DataTable's inline. Something like:

```dart
const List<DataColumn> columns = <DataColumn>[
  DataColumn(label: Text('column1')),
  DataColumn(label: Text('column2')),  
];

const List<DataCell> cells = <DataCell>[
  DataCell(Text('cell1')),
  DataCell(Text('cell2')),  
];

const List<DataRow> rows = <DataRow>[
  DataRow(cells: cells),
  DataRow(cells: cells),  
];
```

And then
```dart
await tester.pumpWidget(
  MaterialApp(
    home: Material(
      child: DataTable(
        rows: rows,
        columns, columns,
      )
    ),
  ),
);
```
etc.

HansMuller commented on packages/flutter/test/material/data_table_test.dart:
```diff
@@ -866,4 +866,89 @@ void main() {
+      home: Material(child: buildTable()),
+    ));
+
+    Table table = tester.widget(find.byType(Table).first);
```
The `.first` shouldn't be needed here and below.

HansMuller commented on packages/flutter/test/material/data_table_test.dart:
```diff
@@ -866,4 +866,89 @@ void main() {
+            );
+          }).toList(),
+        );
+      }else {
```
`} else {`

HansMuller reviewed (commented):
This is looking good, just some small stuff.

dratushnyy committed:
Material DataTabel: renamed border thickness constructor parameter and
field from 'borderThickness' to 'dividerThickness'.
Refactored related unit tests: simplify the creation of test data
table

Piinks commented on packages/flutter/lib/src/material/data_table.dart:
```diff
@@ -467,6 +469,11 @@ class DataTable extends StatelessWidget {
   static const Color _grey300Opacity = Color(0x1E000000); // Dark theme variant is just a guess.
 
+  /// The width of the divider that appears between [TableRow]s.
+  /// Must be non-null and greater than or equal to zero.
```
```suggestion
  /// 
  /// Must be non-null and greater than or equal to zero.
```

Nit: Dart docs specify a blank line should follow each paragraph, and that the first sentence should be it own paragraph.

Piinks reviewed (approved):
This LGTM @dratushnyy, with one tiny nit below. Thank you for the contribution.
It also looks like your branch contains a flake that is causing the `customer_*` checks to fail. This is unrelated to your change, and should be resolved if you update your branch with changes from master. :)

dratushnyy committed:
Reformatted documentation strings for "dividerThickness".

Co-Authored-By: Kate Lovett <katelovett@google.com>

HansMuller commented:
Looks like a trival trailing whitespace error needs to be fixed:
```
packages/flutter/lib/src/material/data_table.dart:473: trailing U+0020 space character
```

dratushnyy committed:
Material Datatable: removed trailing space in documentaion
strings for `dividerThinkess` property

dratushnyy commented:
@HansMuller fixed. Sorry for the formatting issues.

Piinks commented:
It looks like this branch still needs to be updated to resolve the failing checks. :)

dratushnyy committed:
Merge remote-tracking branch 'upstream/master' into data_table_set_border_thikness

Piinks commented:
Thanks for the update! :)
